### PR TITLE
Fix #7590: Allow 'readonly' to be used in constructor parameters

### LINKED ISSUE #7590: Suggestion: readonly constructor parameters
I've been using the new `readonly` properties feature (#6532) and have found it to be a very productive alternative to using getters or `getWhatever()` functions, or as is often the case, simply doing nothing to enforce read-only properties. However I'm really yearning for the ability to mark constructor parameters as `readonly`.

In 1.8 a common pattern for me is to write classes where all properties are public and readonly like so:

```typescript
export class Foo {
  constructor(public bar: number,
              public baz: number) {}
}

```

But now with `readonly` I'm missing out unless I rewrite the above code to:

```typescript
export class Foo {
  public readonly bar: number;
  public readonly baz: number;

  constructor(bar: number, baz: number) {
    this.bar = bar;
    this.baz = baz;
  }
}
```

Which is pretty cumbersome. What I'd really like to be able to write is:

```typescript
export class Foo {
  constructor(public readonly bar: number,
              public readonly baz: number) {}
}
```

Right now I'm worried about hitting 2.0 and having to rewrite all my constructors just to take advantage of `readonly`. Because it's an advantage I definitely want to have.

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #7590

### CODE REVIEW

File: src/compiler/types.ts [hunk 1]
```diff
@@ -407,8 +407,10 @@ namespace ts {
         HasAggregatedChildData = 1 << 29,  // If we've computed data from children and cached it in this node
         HasJsxSpreadAttribute = 1 << 30,
 
-        Modifier = Export | Ambient | Public | Private | Protected | Static | Abstract | Default | Async,
+        Modifier = Export | Ambient | Public | Private | Protected | Static | Abstract | Default | Async | Readonly,
         AccessibilityModifier = Public | Private | Protected,
+        // Accessibility modifiers and 'readonly' can be attached to a parameter in a constructor to make it a property.
```
Comment (DanielRosenwasser) [inline src/compiler/types.ts]: The term for this is a "parameter property", so maybe a better name for the flag would be `ParameterPropertyModifier`

File: src/compiler/types.ts [hunk 2]
```diff
@@ -407,8 +407,10 @@ namespace ts {
         HasAggregatedChildData = 1 << 29,  // If we've computed data from children and cached it in this node
         HasJsxSpreadAttribute = 1 << 30,
 
-        Modifier = Export | Ambient | Public | Private | Protected | Static | Abstract | Default | Async,
+        Modifier = Export | Ambient | Public | Private | Protected | Static | Abstract | Default | Async | Readonly,
```
Comment (yuit) [inline src/compiler/types.ts]: Do you still want to keep `Readonly` in ? since you have made a new flag `ParameterPropertyModifier`

**Comment by yuit** [comment]
I believe that TypeScript [Spec](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#831-constructor-parameters) should be updated to include this change

File: tests/cases/conformance/classes/constructorDeclarations/constructorParameters/readonlyInConstructorParameters.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+class C {
+    constructor(readonly x: number) {}
```
Comment (yuit) [inline tests/cases/conformance/classes/constructorDeclarations/constructorParameters/readonlyInConstructorParameters.ts]: also to clarify if it is just `readonly` without accessibility modifier, are we assuming it is `public`?

Comment (andy-ms) [inline tests/cases/conformance/classes/constructorDeclarations/constructorParameters/readonlyInConstructorParameters.ts]: Yep. This is the same as for properties. (`class C { readonly x: number; }`)

File: src/compiler/types.ts (hunk 2 above)
Comment (mhegazy) [inline src/compiler/types.ts]: it is a modifier..

**Comment by mhegazy** [comment]
Can you add a test for declaration emitter. i.e. `// @declaration: true` at the top of the file. see https://github.com/Microsoft/TypeScript/blob/master/tests/cases/compiler/declarationEmit_protectedMembers.ts for an example.

**Comment by mhegazy** [comment]
Also a test for:
- readonly in method signature and method declaration
- readonly in setter declarations
- lambda function with a parameter named `readonly`.

File: src/compiler/types.ts (hunk 2 above)
Comment (yuit) [inline src/compiler/types.ts]: got cha, I got confused why it is here caz i thought it is not direct related with the bug.

**Comment by mhegazy** [comment]
Please add the word doc updates to the spec as well.

**Comment by mhegazy** [comment]
:+1:

**Comment by DanielRosenwasser** [comment]
I think it'd be better to just file a spec bug and assign it to me or Anders .

**Pull request merged**